### DISCUSSION TITLE
Is there a Rule trigger for when IP address renewed via DHCP

### DISCUSSION THREAD

**Discussion by murfett-au** [arendst/Tasmota#19513] [Q&A] [answered]
Is there a rule trigger that will run a rule every time a device gets a new IP addres from a DHCP server?

I like to be able to quickly find the IP addresses of my tasmota devices so I can access their web UI quickly.  My devices are all set up with MQTT, so I can subscribe to the `tele/+/INFO2` in NodeRed, and publish (or "re-broadcast") the IP to an `ipaddresses/byHostname/+` topic. By setting the retain flag to true for that message, I can start my MQTT client and immediately get a list of ip addresses by hostname.

This executes on system boot, but NOT after a new DHCP address is assigned by my DHCP server.

This is the code I use in the nodered funtion node - I could achieve the same result many other ways (eg mosquitto plug in running on the broker) - my question is not about this code, but I am including it for context.
```
const payload = JSON.parse(msg.payload);
const hostname = payload.Info2.Hostname;
msg.payload = `${payload.Info2.IPAddress} ${new Date().toISOString()}`;
msg.topic=`ipaddresses/byHostname/${hostname}`;
return msg;
```

**Comment by s-hadinger** [comment]
IPv4 addresses typically don't change when they are assigned, even after a reboot. Unless your dhcp is doing weird things. 

I'm not sure what is your problem. 

Also bear in mind that a Tasmota device can have up to 6 IP addresses. Ipv4 for wifi and ethernet, and 2x2 ipv6

**Reply by murfett-au** [reply]
I agree it is strange that the IP addresses change, but they do. I am using a DLink Deco Mesh WiFi.

Thanks for your comments - and my question is still valid.

To clarify I am looking specifically at the WiFi V4 address. Thanks for pointing out there are multiple IPs per device.

**Reply by sfromis** [reply]
When you say "they do", does that mean that you see Tasmota being operational on a changed IP address suddenly, without having restarted?

Or does it only mean that the your mesh would "like" to provide a different IP upon next DHCP request?

**Reply by murfett-au** [reply]
My Tasmota devices get assigned a different IP address by the DHCP server when they renew their leases, without restarting. When this happens I can scan the subnet looking for the new IP or I can reboot it (publish "1" to cmnd/device/restart, or power cycle) and it will publish the info packet so I know the new IP.

I agree it's weird.

**Reply by s-hadinger** [reply]
I have never seen this, and it's hard to test because "normal" DHCP would happily reassign the same IP when the renewal happens. I'm afraid it's hard to help you here.

What is your lease duration? And why on earth would the DHCP server refuse to renew the current address?

**Reply by sfromis** [reply]
Do you *actually* see the new IP address *operational* in Tasmota itself, when checking IP from within Tasmota? Or only based on what the DHCP tells you?

What I mean is that it might have decided on a new IP for "next lease", and already route traffic for the new IP to the old one in actual use.

**Reply by barbudor** [reply, marked as answer]
I don't think there is an event when it changes but you could use rules to periodically issue an IpAddress1 command, so that your backoffice will know about any change

> When this happens I can scan the subnet looking for the new IP or I can reboot it (publish "1" to cmnd/device/restart, or power cycle) and it will publish the info packet so I know the new IP.

If I understand your description well, an new IP is assigned without reboot but then if you reboot you will get hte same IP as before

**Reply by murfett-au** [reply]
I confirm the IP address has changed by trying to hit the old IP in chrome, and the Web UI does not show up. I then visit the web UI of my WiFi base station and look at the list of IPs it has assigned, and find a new IP address. When I hit this IP address, the Tasmota Web UI shows up as I expect. Neither the WiFi nor the tasmota has power cycled when this has happened. If I enter IPADDRESS in the console, the tasmota tells me it is using the new IP address.

I agree that RFC2131 says a DHCPREQUEST should be used to renew a lease, and DHCPDISCOVER should only be used initially to get an IP address (not sure if it is request or discover when rebooting, and not sure if the server just gives the same mac address the same address, or if the client includes it's old IP in the REQUEST packet on reboot), but that is not what my question is about. My unusual but real world experience is that these IP addresses are changing. Never seen this before on any other DHCP server. Unfortunately I cannot turn off the DHCP server on this tplink deco WiFi either.

I might look at the DHCP server logs next time I see this happening to see what it is doing.

So it sounds like there is no rule trigger, which is fine.

I appreciate @barbudor 's advice and may add a RuleTimer triggered rule to do IpAddress1 periodically.

**Reply by sfromis** [reply]
So you did not confirm that Tasmota as such is seeing and using a different IP address *without having restarted*. I did not ask about "power cycled", but restarts (or reconnects in general) for whatever reason. This obviously also includes Wifi dropping out and coming back. In case of Wifi reconnection, you do have a trigger like `Wifi#Connected` allowing you check current state of IP address.

**Reply by barbudor** [reply]
You may avoid using a rule timer and use teleperiod to trigger for example `on tele-uptime#data do backlog IPAddress 1 endon`
